src/crs.rs

```diff
@@ -1,146 +1,91 @@
 const G1_START: usize = 28;
 const G2_START: usize = 28 + (5_040_001 * 64);
 const G2_END: usize = G2_START + 128 - 1;
 
-const BACKEND_IDENTIFIER: &str = "acvm-backend-barretenberg";
-const TRANSCRIPT_NAME: &str = "transcript00.dat";
+// TODO(#162): Allow downloading from more than just the first transcript
 const TRANSCRIPT_URL: &str =
     "http://aztec-ignition.s3.amazonaws.com/MAIN%20IGNITION/monomial/transcript00.dat";
 
-fn transcript_location() -> PathBuf {
-    match env::var("BARRETENBERG_TRANSCRIPT") {
-        Ok(dir) => PathBuf::from(dir),
-        Err(_) => dirs::home_dir()
-            .unwrap()
-            .join(".nargo")
-            .join("backends")
-            .join(BACKEND_IDENTIFIER)
-            .join(TRANSCRIPT_NAME),
-    }
-}
-
 #[allow(clippy::upper_case_acronyms)]
-#[derive(Clone, Default)]
+#[derive(Clone, Default, Serialize, Deserialize)]
 pub(crate) struct CRS {
     pub(crate) g1_data: Vec<u8>,
     pub(crate) g2_data: Vec<u8>,
-    // num_points: usize,
+    pub(crate) num_points: usize,
 }
 
 impl CRS {
-    pub(crate) fn new(num_points: usize) -> CRS {
+    pub(crate) async fn update(&mut self, num_points: usize) -> Result<(), Error> {
+        // We already have some data, so start at the end of our list
+        let g1_start = self.g1_data.len();
         // UltraPlonk requires a CRS equal to circuit size plus one!
         // We need to bump our polynomial degrees by 1 to handle zero knowledge
         let g1_end = G1_START + ((num_points + 1) * 64) - 1;
 
-        // If the CRS does not exist, then download it from S3
-        if !transcript_location().exists() {
-            download_crs(transcript_location()).unwrap();
-        }
-
-        // Read CRS, if it's incomplete, download it
-        let mut crs = read_crs(transcript_location());
-        if crs.len() < G2_END + 1 {
-            download_crs(transcript_location()).unwrap();
-            crs = read_crs(transcript_location());
-        }
-
-        let g1_data = crs[G1_START..=g1_end].to_vec();
-        let g2_data = crs[G2_START..=G2_END].to_vec();
-
-        CRS {
-            g1_data,
-            g2_data,
-            // num_points,
-        }
-    }
-}
-
-// TODO(blaine): Come up with a better abstraction for the CRS so we don't need to read the
-// file everytime we need the G2
-pub(crate) struct G2 {
-    pub(crate) data: Vec<u8>,
-}
-
-impl G2 {
-    pub(crate) fn new() -> G2 {
-        // If the CRS does not exist, then download it from S3
-        if !transcript_location().exists() {
-            download_crs(transcript_location()).unwrap();
-        }
-
-        // Read CRS, if it's incomplete, download it
-        let mut crs = read_crs(transcript_location());
-        if crs.len() < G2_END + 1 {
-            download_crs(transcript_location()).unwrap();
-            crs = read_crs(transcript_location());
-        }
+        // TODO(blaine): Make sure g1_start isn't off-by-one
```

Comment:
Could add a test where `download_crs(M)?.update(N) == download_crs(M+N)`? (For small values of M/N so that we're not downloading 300MB)

Reply:
Yeah, that's a good idea. We'll need to check it against a constraint system, but we can start small and then give a larger constraint system.